### ISSUE TITLE
Reg\Transfers - Internal User can transfer a facility to another operation

### ISSUE THREAD

**Issue by andrea-williams** [bcgov/cas-registration#2427]
#### Dependencies

- Blocked by #1432 
- Blocking #1949 

#### Description:

As an internal user, I need to be able to transfer a facility to another operation

#### Acceptance Criteria:
Given that I am on the [BCIERS dashboard](https://www.figma.com/design/LsDmLDCdnJqI8UahyqvOD8/BCIERS-Designs?node-id=7781-6471&t=riHU6HU0lkTtoc8R-4) and click "Transfer an Operation or Facility"
OR Given that I am on the [Transfer Grid,](https://www.figma.com/design/LsDmLDCdnJqI8UahyqvOD8/BCIERS-Designs?node-id=23653-2739&t=DXPcYyBQv3MoCu6v-4) and I click "Make a transfer"
Then I am brought to the Transfer Entity form [Figma](https://www.figma.com/design/LsDmLDCdnJqI8UahyqvOD8/BCIERS-Designs?node-id=23653-2740&t=DXPcYyBQv3MoCu6v-4)

Given I haven't yet answered the question "What is being transferred?"
The only fields that appear on the page are "Current operator", "Select the new operator", and "What is being transferred?"
And the "Transfer Entity" button is disabled

Given that I have selected "Facility" as my answer to the "What is being transferred?" question,
Then only the facility-specific fields of the form are displayed ("Select the operation that this facility currently belongs to", "Facility" dropdown, "Select the new operation this facility will be allocated to", and "Effective date of transfer")

Given that I submitted the form,
Then I am brought to the transfer complete success page
[Figma](https://www.figma.com/design/LsDmLDCdnJqI8UahyqvOD8/BCIERS-Designs?node-id=23653-2840&t=h6dc4NPJ0ohAfkK5-4)

**Context: although the wireframe shows fields for transferring an operation and a facility, this ticket only captures the flow for transferring a facility. The work for transferring an operation is captured in separate ticket #1949**

**Context: this ticket only covers CREATING a new transfer record for facilities. Editing the transfer form will be covered in a separate ticket.**

Additional Context: see associated ticket #2040 

#### Development Checklist:

- [x] apply RBAC to API endpoint(s) for making a transfer so that it's only accessible to "cas_admin" and "cas_analyst" users
- [x] in "Create" mode of a new Transfer, the breadcrumb [Name of entity being transferred] (in wireframe) does NOT need to populated - developer's choice to leave it entity or add generic title like "Transfer Entity"/"New Transfer"
- [ ] ~the dropdown options in "Select the new operator" dropdown should exclude the operator selected as "Current operator"~
- [x] the options in "Select the operation that this facility currently belongs to" dropdown should only be the operations associated with the Current Operator (selected in a previous dropdown)
- [x] the options in the "Facility" dropdown should only be the facilities that are associated with the selected operation (from the previous dropdown)
- [x] the options in the "Select the new operation this facility will be allocated to" dropdown should only be the operations associated with the "Select the new operator" dropdown answer 
- [ ] ~for each facility listed in the Transfer Entity form, a new record should be created in the `transfer_event` table~ => **Sep**: This is not valid, we are saving facilities as a M2M field
- [x] for each facility listed in the Transfer Entity form, a new record should be created in `facility_designated_operation_timeline` table with the form fields. The "effective date of transfer" in the form should be inserted into the `start_date` column for the new `FacilityDesignatedOperationTimeline` record, and it should also be inserted into the `end_date` column for the `FacilityDesignatedOperationTimeline` record of the previous designated operation. The `status` of the new record should be set to `To be transferred`. Transferring a facility should not change any data in the `facility` table => **Sep**: After processing the transfer, we create an entry in the `FacilityDesignatedOperationTimeline` table. (we don't do it if transfer is in the future)
- [x] create several mock fixtures for transfers of facilities, with effective dates both in the distant future and in the past
- [x] recommended: add explanation in `docs/backend` folder to explain how transfers work - it's complicated and confusing
- [ ] Meets the DOD

**Definition of Ready** (Note: If any of these points are not applicable, mark N/A)

- [ ] User story is included
- [ ] User role and type are identified
- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if required)
- [ ] Design / Solution is accepted by Product Owner
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

·**Definition of Done** (Note: If any of these points are not applicable, mark N/A)

- [ ] Acceptance criteria are tested by the CI pipeline
- [ ] UI meets accessibility requirements
- [ ] Configuration changes are documented, documentation and designs are updated
- [ ] Passes code peer-review
- [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
- [ ] Ticket is ready to be merged to main branch
- [ ] Can be demoed in Sprint Review
- [ ] Bugs or future work cards are identified and created
- [ ] Reviewed and approved by Product Owner

**Comment by Sepehr-Sobhani** [comment]
Questions and Comments:
- What is the function of the “Add Another Facility” button?
- Should we only display operator-related drop-downs if the user intends to transfer an operation?
- Why do we need to ask the user to select the operation that this facility belongs to? We already have that information. 🤔
- I recommend addressing the issue of changing the status of the second data center from “PENDING” to “To Be Transferred” in the transfer data grid ticket.
- We have a note indicating that the facility drop-down is multi-select. Does this mean that all selected facilities will be transferred to a single operation?

**Comment by andrea-williams** [comment]
- "Add Another Facility" button is for when multiple facilities from the same operation need to be transferred at the same time to the same recipient operation
- I'm not sure I understand your question, but facilities could be transferred to a different operator too, so I think we'll need to ask about the operator regardless
- We would be asking them to select the operation that the facility currently belongs to just so that the facility dropdown then only shows the facilities for the selected operation.
- I'm on board for that
- yes, all selected facilities would be transferred to the one recipient operation selected

**Comment by andrea-williams** [comment]
@zoeyli-46 design question (all credit goes to @Sepehr-Sobhani for this idea): rather than having the "Add another facility" button and one dropdown per facility, could we instead make the "Facility" dropdown a multi-select, and then include the geographic coordinates in the dropdown options. (So for example, each option in the facility dropdown would look like "[Operation name] - [lat, long]"). We think this way would be easier to implement development-wise, and it would also save space on the page.

**Comment by Sepehr-Sobhani** [comment]
We should change the wording of 'transfer completion success page' because transfers have an effective date, and we might not transfer entities immediately.
cc: @andrea-williams / @zoeyli-46 / @patriciarussellCAS

**Comment by Sepehr-Sobhani** [comment]
Hi @patriciarussellCAS / @andrea-williams!

For transferring facilities, does the user want to transfer a facility from Operation 1 to Operation 2 under the same operator, or should facilities be transferable from Operation 1 under Operator 1 to Operation 2 under Operator 2, or both?

**Comment by andrea-williams** [comment]
> Hi [@patriciarussellCAS](https://github.com/patriciarussellCAS) / [@andrea-williams](https://github.com/andrea-williams)!
> 
> For transferring facilities, does the user want to transfer a facility from Operation 1 to Operation 2 under the same operator, or should facilities be transferable from Operation 1 under Operator 1 to Operation 2 under Operator 2, or both?

I believe both

**Comment by Sepehr-Sobhani** [comment]
> > Hi [@patriciarussellCAS](https://github.com/patriciarussellCAS) / [@andrea-williams](https://github.com/andrea-williams)!
> > For transferring facilities, does the user want to transfer a facility from Operation 1 to Operation 2 under the same operator, or should facilities be transferable from Operation 1 under Operator 1 to Operation 2 under Operator 2, or both?
> 
> I believe both

In this case, we should allow the user to select the same operator for both the current and new operator. We should remove below AC:
> the dropdown options in "Select the new operator" dropdown should exclude the operator selected as "Current operator"

**Comment by andrea-williams** [comment]
> > > Hi [@patriciarussellCAS](https://github.com/patriciarussellCAS) / [@andrea-williams](https://github.com/andrea-williams)!
> > > For transferring facilities, does the user want to transfer a facility from Operation 1 to Operation 2 under the same operator, or should facilities be transferable from Operation 1 under Operator 1 to Operation 2 under Operator 2, or both?
> > 
> > 
> > I believe both
> 
> In this case, we should allow the user to select the same operator for both the current and new operator. We should remove below AC:
> 
> > the dropdown options in "Select the new operator" dropdown should exclude the operator selected as "Current operator"

I agree. I've removed that AC :)

**Comment by patriciarussellCAS** [comment]
Thanks guys! Yes, it could be both.

**Comment by patriciarussellCAS** [comment]
Review notes: 
Transfer an Operation:
First time I tried I got the following error message

![Image](https://github.com/user-attachments/assets/56ce3441-84c1-49e2-85bd-d10bc098c7ab)

The second time I tried, there are no operations available to select in the dropdown. This happens when I select transfer facility as well. 

![Image](https://github.com/user-attachments/assets/f58ee3a4-efb3-47e9-895c-c13b76445fd5)

cc: @Sepehr-Sobhani @BCerki @andrea-williams

**Comment by BCerki** [comment]
> Review notes: Transfer an Operation: First time I tried I got the following error message
> 
Mock data again. In the `Operation` model, Operation 17 belongs to Existing Operator 2 Legal Name, but in the `OperationDesignatedOperatorTimeline`, it belongs to and is active with `Operator 1 Legal Name`. `_process_operation_transfer` should set the end_date for the timeline table (`if current_timeline:
            OperationDesignatedOperatorTimelineService.set_timeline_status_and_end_date(...`) but we don't find a current_timeline because the operator is wrong. So then there's no end date and we hit the constraint about not having an operation owned by two operators at the same time

> 
> The second time I tried, there are no operations available to select in the dropdown. This happens when I select transfer facility as well.
> 
The dropdowns only show operations that belong to the selected operator, and facilities that belong to the selected operation. New Operator 11 doesn't have any operations, so that's why you're getting nothing in the dropdown. Existing Operator 2 has lots of operations and facilities if you'd like to play around with it, but I'm also going to update the mock data and then everything should work better
> 
> cc: [@Sepehr-Sobhani](https://github.com/Sepehr-Sobhani) [@BCerki](https://github.com/BCerki) [@andrea-williams](https://github.com/andrea-williams)

**Comment by zoeyli-46** [comment]
<img width="1254" alt="Image" src="https://github.com/user-attachments/assets/dfe34080-3925-4ed4-a6e2-9946a18666ca" />

Is this the format for displaying "time" across the app? I wonder if it would be easier to read if it just said "Jul 5, 2024 4:25PM". is the time zone and second necessary information? @patriciarussellCAS

**Comment by zoeyli-46** [comment]
@patriciarussellCAS Tiegan and Kirk confirmed they do not need the time for report submissions and transfers reports. [Thread](https://teams.microsoft.com/l/message/19:143b7487e8a34e28b0660f9e39d8d502@thread.tacv2/1734743257794?tenantId=6fdb5200-3d0d-4a8a-b036-d3685e359adc&groupId=29821293-3244-455d-beae-e01d27762cfe&parentMessageId=1734743257794&teamName=External%3A%20CAS%20-%20Clean%20Growth%20Digital%20Service%20Team&channelName=Carbon%20Pricing%20Transition%20SMEs&createdTime=1734743257794&ngc=true) for context

**Comment by zoeyli-46** [comment]
Created ticket #2640 regarding the date/time display @patriciarussellCAS cc @vesselak
